src/pdf_parser/pdf_utils/ocr.py

```diff
@@ -0,0 +1,487 @@
+            union = google_block.union(lp_block).area
+            dd_intersection_over_union[ix_goog].append(intersection / union)
+
+    # If a google block is covered by a layoutparser block by more than 0.9, we can assume that google OCR has
+    # identified the same block as layoutparser + heuristics and we can use the text from the google block.
+    # Filter the default dict for these cases
+    equivalent_block_mapping = {
+        k: v.index(max(v))
+        for k, v in dd_intersection_over_union.items()
+        if max(v) > threshold
+    }
+
+    # New blocks to add to the layoutparser layout
+    blocks_google_only = {
+        k: v for k, v in dd_intersection_over_union.items() if max(v) == 0.0
+    }
+
+    # use the mapping to replace the text of the layoutparser block with the text of the google block
+    for k, v in equivalent_block_mapping.items():
+        google_coords = google_layout[k].coordinates.vertices
+        (
+            x_top_left,
+            y_top_left,
+            x_bottom_right,
+            y_bottom_right,
+        ) = google_coords_to_lp_coords(google_coords)
+        lp_layout[v].text = google_layout[k].text
+        # TODO: This is ugly. Should create a data type to make these changes more explicit/to not duplicate code
+        lp_layout[v].language = google_layout[k].language
+        lp_layout[v].confidence = google_layout[k].confidence
+        # Reset the coordinates of the layoutparser block to the coordinates of the google block.
+        lp_layout[v].block.x_1 = x_top_left
+        lp_layout[v].block.y_1 = y_top_left
+        lp_layout[v].block.x_2 = x_bottom_right
+        lp_layout[v].block.y_2 = y_bottom_right
+
+    # Add the blocks that are only in the google layout, but only if they are not too small or are too high
+    # up/down on the page indicating that they are probably not part of the main text.
+    for key, val in blocks_google_only.items():
+        google_coords = google_layout[key].coordinates.vertices
+        (
+            x_top_left,
+            y_top_left,
+            x_bottom_right,
+            y_bottom_right,
+        ) = google_coords_to_lp_coords(google_coords)
+        if (
+            y_top_left > image.height * bottom_exclude
+            or y_bottom_right < image.height * top_exclude
+        ):
+            rect = Rectangle(
+                x_1=x_top_left, y_1=y_top_left, x_2=x_bottom_right, y_2=y_bottom_right
+            )
+            lp_layout.append(
+                TextBlock(
+                    rect,
+                    text=google_layout[key].text,
+                    type="Google Text Block",
+                    score=google_layout[key].confidence,
+                )
+            )
+            # TODO: This is ugly. Should create a data type to make these changes more explicit/to not duplicate code
+            lp_layout[-1].language = google_layout[key].language
+
+    return lp_layout
+
+
```

Comment:
I retained the OCRProcessor class. Questionable whether we want to keep this as a class, but we still need the old methodology in case LayoutParser detects some boxes that google doesn't and so we don't have the text from a page level API call. This will still reduce costs significantly (if we only need to do additional API calls for 1 box per page, say).